urdf/ur_macro.xacro

```diff
@@ -129,126 +148,71 @@
     <!-- Placeholder for ros2_control transmission which don't yet exist -->
 
     <!-- links -  main serial chain -->
+	<!-- Special exception in creation is the base_link_inertia, as naming convention is slightly different from other links which use xacro:new_link. -->
     <link name="${prefix}base_link"/>
     <link name="${prefix}base_link_inertia">
       <visual>
         <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
         <geometry>
-          <mesh filename="${base_visual_mesh}"/>
+          <xacro:get_mesh name="base" type="visual"/>
         </geometry>
       </visual>
       <collision>
         <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
         <geometry>
-          <mesh filename="${base_collision_mesh}"/>
+          <xacro:get_mesh name="base" type="collision"/>
         </geometry>
       </collision>
       <xacro:cylinder_inertial radius="${base_inertia_radius}" length="${base_inertia_length}" mass="${base_mass}">
         <origin xyz="0 0 0" rpy="0 0 0" />
       </xacro:cylinder_inertial>
     </link>
-    <link name="${prefix}shoulder_link">
-      <visual>
-        <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
-        <geometry>
-          <mesh filename="${shoulder_visual_mesh}"/>
-        </geometry>
-      </visual>
-      <collision>
-        <origin xyz="0 0 0" rpy="0 0 ${pi}"/>
-        <geometry>
-          <mesh filename="${shoulder_collision_mesh}"/>
-        </geometry>
-      </collision>
+
+    <xacro:new_link name="shoulder">
```

Comment:
This looks nice, but the question is if the code becomes really more readable using this. It is not very common, so it could also be confusing to many people.

Reply:
They way I see it, it's the first step of a bigger refactoring.

Something I would aim for if I would work more on it would be one xacro with an input of a list with all the required info, that loops through and creates all the joints and links.

As I said in the first comment, it looks like there is a lot that can be refactored into smaller functions, etc etc.
I think that would make it more readable at the same time too. Though ofc it's never perfect, it could always be easier :)

Reply:
That is also, I guess, a matter of where you come from. I find the pure URDF really hard get a full understanding of, it's just too much, I guess I'm a lazy programmer 😃 

But yes, it's hiding information, information that is repeating over and over again, adding a layer of abstraction. Personally, I find it much easier to read small abstracted functions that hide repetitive information, than big models. Also I find that most things are already hidden quite deep, but these kinds of refactorings could reduce the layers but keep (or as I would prefer it, increase) the level of abstraction

I do think though that the URDFs could be much easier generated with an OOP language, than Xacro, as I said, though I guess that would require more work.

Reply:
> I guess I'm a [..] programmer

And I guess that's the reason why you like doing the things you did in this PR.

Many users of ROS (and these models) are not programmers.

Every abstraction and every layer of indirection adds complexity.

In my experience with these kinds of things, and the kinds of users these packages see, more layers and abstraction do not equal better understanding, or better user experiences.

That's the reason things are "repeating over and over again" in some cases: it shows what's going on, exactly as you see, where you see it. Yes, it's repeated, but it's a small price to pay for an often acceptable trade-off.

Looking up a `xacro:macro` definition and mentally computing what's happening may be second nature to you (as it would be for me), but there are sufficient users for which this is not the case.

Whether the maintainers here agree or not I don't know, but I wanted to leave this perspective here, based on my experience.

Reply:
True, and you are right, I absolutely agree that this way is not for everyone. But I think it is already very complicated, already mixing macros and hardcoding, which I think is more confusing than one or the other.

Personally I think the best tradeoff would be a refactored abstracted xacro, which the users could easily with a provided script build into a URDF to edit by hand if they find the "programmers way" more complicated.